### PR TITLE
router: add more stats to virtual clusters

### LINKED ISSUE #10449: virtual clusters: emit additional metrics
*Description*:
Currently [Envoy Mobile](https://github.com/lyft/envoy-mobile) uses (effectively) a single dynamic forward proxy cluster for all of its traffic. This results in dramatically simpler ergonomics for users of the library. However, it reduces the granularity of metrics available because all traffic is going through the same cluster.

This is a perfect use case for [Virtual Clusters](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route_components.proto#envoy-api-msg-route-virtualcluster), specially since commonly users are only interested in granular traffic for a small percentage of requests. Hence that percentage can be segmented as desired using virtual clusters.

Currently virtual clusters only expose:
```
upstream_rq_<*xx> - Counter - Aggregate HTTP response codes (e.g., 2xx, 3xx, etc.)
upstream_rq_<*> - Counter - Specific HTTP response codes (e.g., 201, 302, etc.)
upstream_rq_time - Histogram - Request time milliseconds
```

This ticket tracks exposing a couple more metrics of interest:
```
upstream_rq_timeout - Counter - Total requests that timed out waiting for a response
upstream_rq_per_try_timeout - Counter - Total requests that hit the per try timeout
upstream_rq_retry - Counter - Total request retries
upstream_rq_retry_success - Counter - Total request retry successes
upstream_rq_total - Counter - Total request
```

Note: relatedly if a user is interested in per vhost, per route stats _for all_ vhosts, routes there is an open issue https://github.com/envoyproxy/envoy/issues/3351

### PR DESCRIPTION
**Pull request by junr03**
Description: adds additional stats emitted by virtual clusters. 
Risk Level: low
Testing: added tests for stats values in existing unit tests that test for functionality that modifies the new stats. Additionally added an integration test to make sure that the stats are getting added to the correct namespace.
Docs Changes: modified docs to add information about the new stats.
Release Notes: added a release note.

Fixes #10449

### CODE REVIEW

**Comment by junr03** [comment]
Creating a draft PR to get an initial opinion from @jmarantz about using strongly named stats with a chain of new Stats::Scopes. My thinking is that these stats are static and not dynamically generated like the rest of the current stats emitted in the virtual cluster namespace, so it didn't make sense to use the symbol table approach.

If you think that makes sense I will go ahead and finish up this PR by adding tests, docs, and release notes.

**Comment by jmarantz** [comment]
I just skimmed through this. I am not sure what you mean by "strongly named" but the main question I have is whether the virtual clusters are created dynamically in the line of requests. If so then I think we can make some small changes to reduce potential contention.

File: include/envoy/router/router.h [hunk 1]
```diff
@@ -370,6 +370,22 @@ class ShadowPolicy {
...
 using ShadowPolicyPtr = std::unique_ptr<ShadowPolicy>;
 
+/**
+ * All virtual cluster stats. @see stats_macro.h
+ */
+#define ALL_VIRTUAL_CLUSTER_STATS(COUNTER)                                                         \
+  COUNTER(upstream_rq_retry)                                                                       \
+  COUNTER(upstream_rq_retry_success)                                                               \
+  COUNTER(upstream_rq_timeout)                                                                     \
+  COUNTER(upstream_rq_total)
```
Comment (jmarantz) [inline include/envoy/router/router.h]: I was mulling over defining another layer of macros so you could pre-declare these names in a config structure or factory, without binding to a scope, and then have a separate macro that instantiates the stats in a scope.

**Comment by junr03** [comment]
I was using historical language (similar to https://github.com/envoyproxy/envoy/blob/master/include/envoy/upstream/upstream.h#L822). Mostly meant stats that are not created dynamically in the data path.

So to answer your question: virtual cluster entries (and thus the stats struct I added here) are only constructed once during server initialization when we parse router filter configs and build the filter config object

https://github.com/envoyproxy/envoy/blob/1db15db60af37b167b72e3c5e653ffdda9711336/source/common/router/config_impl.cc#L1001

**Comment by jmarantz** [comment]
ok then this approach sgtm :)

**Comment by jmarantz** [comment]
ping when ready for review 

/wait

**Comment by junr03** [comment]
@jmarantz ready for a pass

File: include/envoy/router/router.h [hunk 2]
```diff
@@ -370,6 +370,22 @@ class ShadowPolicy {
 
 using ShadowPolicyPtr = std::unique_ptr<ShadowPolicy>;
 
+/**
+ * All virtual cluster stats. @see stats_macro.h
+ */
+#define ALL_VIRTUAL_CLUSTER_STATS(COUNTER)                                                         \
+  COUNTER(upstream_rq_retry)                                                                       \
```
Comment (jmarantz) [inline include/envoy/router/router.h]: should we have tests for the other counters? I just see tests for 2.

Comment (junr03) [reply include/envoy/router/router.h]: They all have tests:
1.`upstream_rq_retry` and `upstream_rq_retry_success`: https://github.com/envoyproxy/envoy/pull/10474/files#diff-0a6a6327b31a217b22c653b1d1ebf4acR829
2.`upstream_rq_timeout`: https://github.com/envoyproxy/envoy/pull/10474/files#diff-6eb802c7b8ea53fcfa4c3f2842b28c7bR1565
3.`upstream_rq_total`: all over router_test.cc

Comment (junr03) [reply include/envoy/router/router.h]: Ah maybe router_test was collapsed in your view?

Comment (jmarantz) [reply include/envoy/router/router.h]: Exactly "large diffs are hidden by default" bit me again. Sorry about the confusion!

**Pull request merged**